### DISCUSSION TITLE
Default Greeting workflow suddenly resulting in an error

### DISCUSSION THREAD

**Discussion by opticSquid** [community/community#25706] [Actions]
<p>My repo uses <strong>Greetings</strong> workflow by <code>Github Actions</code> which greets first time contributors to a repo with custom messages when they do a PR or raise an issue. The workflow was working fine for issues but when someone tried to create a PR from <code>their fork</code> to the <code>main</code> branch of the main repo the workflow shows error.</p>
<p>My workflow is:</p>
<pre><code class="lang-auto">name: Greetings

on: [pull_request, issues]

jobs:
  greeting:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
    - uses: actions/first-interaction@v1
      with:
        repo-token: ${{ secrets.GITHUB_TOKEN }}
        issue-message: 'Message that will be displayed on users first issue'
        pr-message: 'Message that will be displayed on users first pull request'

</code></pre>
<p>The error it shows:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/d/7/d7f78e36b8d3f93bcaf60ffeacb9936838501812.png" data-download-href="/uploads/short-url/uOx2zaxg6eSKYSlPg2PsFVLXZES.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/59192460/181092748-c179168b-2c1d-45f7-8da8-9326b1abb56c.png" alt="image" data-base62-sha1="uOx2zaxg6eSKYSlPg2PsFVLXZES" width="690" height="316" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/7/d7f78e36b8d3f93bcaf60ffeacb9936838501812_2_690x316.png, https://aws1.discourse-cdn.com/github/optimized/3X/d/7/d7f78e36b8d3f93bcaf60ffeacb9936838501812_2_1035x474.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/d/7/d7f78e36b8d3f93bcaf60ffeacb9936838501812_2_1380x632.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/7/d7f78e36b8d3f93bcaf60ffeacb9936838501812_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1814×832 59.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by airtower-luna** [comment]
<p>The <code>GITHUB_TOKEN</code> for <code>pull_request</code> events <a href="https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token">has read-only permissions</a>. Otherwise people could write malicious workflows, create a PR for your repository, and let those workflows run with write access to your repository.</p>
<p>In your case the <a href="https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target"><code>pull_request_target</code> event</a> might work, which runs a workflow defined on the <em>target</em> branch of a PR (so not modifiable by an attacker without the PR getting merged). Using it to post a comment should be safe, but when writing the workflow you need to make sure it doesn’t allow script injection or running code from the PR, see:</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/5814919/181092750-7aa85c7f-344f-4362-bf30-ce450014a972.png" class="site-icon" width="72" height="72">

      <a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener" title="12:00AM - 15 December 2020">GitHub Security Lab – 15 Dec 20</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/362;"><img src="https://user-images.githubusercontent.com/5814919/181092751-2eeda580-a532-413e-b144-285b40d2ace8.png" class="thumbnail" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_690x362.png, https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_1035x543.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/8/4/844607e16e2775a69e8cd3cfefc2ab35ac88a928_2_10x10.png"></div>

<h3><a href="https://securitylab.github.com/research/github-actions-preventing-pwn-requests/" target="_blank" rel="noopener">Keeping your GitHub Actions and workflows secure Part 1: Preventing pwn requests</a></h3>

  <p>In this article, we’ll discuss some common security malpractices for GitHub Actions and workflows, and how to best avoid them. Our examples are based on real-world GitHub workflow implementation vulnerabilities the GitHub Security Lab has reported to...</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by opticSquid** [comment]
<p>I did not get really what you said it went over my head. Please tell me what modifications I have to do in my workflow to make it work</p>

**Comment by opticSquid** [comment]
<p>Mine is a public repo, one of my contributors have raised a PR from their fork and I’m trying to merge it this greetings work flow doesn’t work thats it</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>I’ll try to simplify it a bit:</p>
<p>Each workflow run automatically gets a thing called a <code>GITHUB_TOKEN</code>. That token is used for authentication, for example when your workflow creates a comment. When the workflow runs because of an issue or PR inside your repository, the token is allowed to make changes.</p>
<p>When a <code>pull_request</code> happens because of a PR from a fork, the token has read-only permissions. It can’t be used to make changes, so you can’t use it to write a comment. This is for security: With the <code>pull_request</code> event the workflow file is taken from the PR. Anyone can make a PR, and someone could edit the workflow file in the PR to do something bad.</p>
<p>There is another event called <code>pull_request_target</code> you can use to make workflows run when a PR is created. The workflow file for <code>pull_request_target</code> must already be in your repository, on the <em>target</em> branch of the PR. So a random PR author can’t change it. The token for those workflows can make changes. However, that also means you must be careful not to run any code from the PR. Otherwise it could abuse the token to do bad things again. The article above is about that problem.</p>

**Comment by opticSquid** [comment]
<p>Thanks for the clarification :blush:</p>
